src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis
```

Comment:
Just thought of while implementing - should this API notify us when a `SourceFileInfo` is removed? (or make it disposable).

Reply:
we can do it if that is necessary but not having Dispose is usually make things easier. do you have a reason why want to get noticed when file is removed? IDocumentServiceProvide being specific to 1 snapshot (Document), not enough for state management?

Reply:
I think I shouldn't say 1 snapshot. I meant 1 green node (DocumentState).